src/libstd/io/buffered.rs

```diff
@@ -15,9 +15,33 @@ use io::prelude::*;
 use cmp;
 use error;
 use fmt;
-use io::{self, DEFAULT_BUF_SIZE, Error, ErrorKind, SeekFrom};
+use io::{self, TrustedRead, DEFAULT_BUF_SIZE, Error, ErrorKind, SeekFrom};
 use memchr;
 
+trait MakeBuffer {
+    fn make_buffer(capacity: usize) -> Vec<u8>;
+}
+
+impl<R> MakeBuffer for R
+    where R: Read
+{
+    default fn make_buffer(capacity: usize) -> Vec<u8> {
+        vec![0; capacity]
+    }
+}
+
+impl<R> MakeBuffer for R
+    where R: TrustedRead
+{
+    fn make_buffer(capacity: usize) -> Vec<u8> {
```

Comment:
FWIW I don't think this can be a free function, I believe specialization only applies to trait impls right now